mypy/checkmember.py

```diff
@@ -370,33 +374,22 @@ def lookup_member_var_or_accessor(info: TypeInfo, name: str,
         return None
 
 
-def check_method_type(functype: FunctionLike, itype: Instance, is_classmethod: bool,
-                      context: Context, msg: MessageBuilder) -> None:
+def check_self_arg(functype: FunctionLike, original_type: Type, is_classmethod: bool,
+                   context: Context, name: str, msg: MessageBuilder) -> None:
+    """Check that the the most precise type of the self argument is compatible
```

Comment:
Style nit: in a multi-line docstring there should be an empty line after the first line.

Reply:
I'll fix it on the next commit or before merge.

Reply:
The docstring could also be clarified a bit. The meaning of "the most precise type of the self argument" wasn't clear initially. Maybe something about generic self, and using the actual type of the self (based on the receiver type) rather than the declared type in the function signature? Would also be good to describe `original_type` here -- it would even make sense to rename it.  What about `actual_self_type`? It's always the type of self, and it's the actual instead of the formal/declared type. If you rename it, it would also make sense to rename it elsewhere.

Reply:
There are 4 potentially distinct types in play here:
1. The declared type of the first parameter `self` (potentially generic)
2. The inferred/declared type of the receiver, ignoring the function signature
3. The same type as (2) but after narrowing it using the current class
4. The type of the current class
(4) is almost irrelevant at this point, but it is used to construct (3), restricts the valid types for (2), and adds to the potential confusion.
So both (2) and (3) are "actual self type". Maybe I should elaborate on the comment, but it will become lengthy.